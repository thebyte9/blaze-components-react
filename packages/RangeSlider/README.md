
## Description

Range slider allows the input of numeric values within a specific range. It can accept a single value, or a range of values (min/max).

## Usage

- With one moveable hander
```js
<RangeSlider
    name="oneValue"
    maxValue={20}
    minValue={0}
    onChange={() => { }}
    value={15} />
```

- With one moveable handler and custom step value
```js
<RangeSlider
    name="step"
    maxValue={20}
    minValue={0}
    onChange={() => { }}
    step={5}
    value={15} />
```

- With two moveable handlers
```js
<RangeSlider
    name="twoValue"
    maxValue={20}
    minValue={0}
    onChange={() => { }}
    value={{ min: 3, max: 7 }} />
```

- With two moveable handlers and draggable track

```js
<RangeSlider
    draggableTrack
    name="draggableTrack"
    maxValue={20}
    minValue={0}
    onChange={() => { }}
    value={{ min: 3, max: 7 }} />
```

## API

##### RangeSlider can receive a number of `props` as follow:


| NAME            | TYPE          | DEFAULT   | 
| :---            | :---:         | :---:     |
| allowSameValues | Boolean       |   false   | 
| disabled        | Boolean       |   false   | 
| draggableTrack  | Boolean       |   false   |  
| maxValue        | Number        |    10     |  
| minValue        | Number        |     0     |  
| name            | String        |     ''    |  
| onChange        | Function      | () => {}  |  
| step            | Number        |     1     |  
| value           | Number/Object |     0     |  
