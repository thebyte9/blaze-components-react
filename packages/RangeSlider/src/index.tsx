import WithUtils from "@blaze-react/utils";
import React, { useState } from "react";
import InputRange from 'react-input-range';

interface IRange {
  max: number;
  min: number;
}

interface IRangeSliderProps {
  allowSameValues?: boolean;
  disabled?: boolean;
  draggableTrack?: boolean;
  maxValue?: number;
  minValue?: number;
  name?: string;
  onChange?: (value: IRange | number) => void
  step?: number;
  value?: IRange | number
}
const RangeSlider: React.SFC<IRangeSliderProps> = ({
  allowSameValues,
  disabled,
  draggableTrack,
  maxValue,
  minValue,
  name,
  onChange,
  step,
  value
}) => {
  const [updatedValue, setUpdatedValue] = useState<IRange | number>(value || 0);

  const handleChange = (newValue: IRange | number) => {
    setUpdatedValue(newValue);
    if (onChange) {
      onChange(newValue);
    }
  };

  return (
    <InputRange
      allowSameValues={allowSameValues}
      disabled={disabled}
      draggableTrack={draggableTrack}
      maxValue={maxValue}
      minValue={minValue}
      name={name}
      onChange={handleChange}
      step={step}
      value={updatedValue}
    />
  );
};
RangeSlider.defaultProps = {
  allowSameValues: false,
  disabled: false,
  draggableTrack: false,
  maxValue: 10,
  minValue: 0,
  name: '',
  onChange: (): void => {
    return;
  },
  step: 1
};
export default WithUtils(RangeSlider);