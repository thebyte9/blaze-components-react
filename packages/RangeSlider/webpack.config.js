const webpackConfig = require('@blaze-react/build-tools/webpack.base.config');

module.exports = webpackConfig;
