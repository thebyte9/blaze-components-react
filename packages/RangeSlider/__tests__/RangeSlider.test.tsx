import { mount } from "enzyme";
import expect from "expect";
import React from "react";
import RangeSlider from "../src";

const defaultProps = (override: object = {}) => ({
  minValue: 0,
  maxValue: 20,
  onChange: () => ({}),
  value: 12,
  ...override
});

describe("RangeSlider component", () => {
  test("should be defined and renders correctly (snapshot)", () => {
    const wrapper = mount(
      <RangeSlider {...defaultProps()} />
    );

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  test("should be disabled", () => {
    const wrapper = mount(
      <RangeSlider {...defaultProps({ disabled: true })} />
    );

    expect(
      wrapper
        .find("InputRange")
        .at(0)
        .prop("disabled")
    ).toBe(true);
  });

  test("should set min, max and value correctly", () => {
    const wrapper = mount(
      <RangeSlider {...defaultProps()} />
    );

    expect(
      wrapper
        .find("span.input-range__label.input-range__label--min span")
        .at(0)
        .text()
    ).toBe("0");

    expect(
      wrapper
        .find("span.input-range__label.input-range__label--value span")
        .at(0)
        .text()
    ).toBe("12");

    expect(
      wrapper
        .find("span.input-range__label.input-range__label--max span")
        .at(0)
        .text()
    ).toBe("20");
  });

  test("should have custom step ", () => {
    const wrapper = mount(
      <RangeSlider {...defaultProps({ step: 4 })} />
    );

    expect(
      wrapper
        .find("InputRange")
        .at(0)
        .prop("step")
    ).toBe(4);
  });

  test("should have draggable track ", () => {
    const wrapper = mount(
      <RangeSlider {...defaultProps({ draggableTrack: true })} />
    );

    expect(
      wrapper
        .find("Track")
        .at(0)
        .prop("draggableTrack")
    ).toBe(true);
  });

  test("should have value with min and max", () => {
    const wrapper = mount(
      <RangeSlider {...defaultProps({ value: { min: 2, max: 8 } })} />
    );

    expect(
      wrapper
        .find("span.input-range__label.input-range__label--value span")
        .at(0)
        .text()
    ).toBe("2");

    expect(
      wrapper
        .find("span.input-range__label.input-range__label--value span")
        .at(1)
        .text()
    ).toBe("8");
  });

  it('should update the current value when one of the arrow keys is pressed', () => {
    const wrapper = mount(
      <RangeSlider
        {...defaultProps({ value: { min: 2, max: 10 }, onChange: (value: any) => wrapper.setProps({ value }) })}
      />
    );

    const slider = wrapper.find(`Slider [onKeyDown]`).first();

    slider.simulate('keyDown', { keyCode: 37 });
    slider.simulate('keyUp', { keyCode: 37 });
    expect(wrapper.props().value).toEqual({ min: 1, max: 10 });

    slider.simulate('keyDown', { keyCode: 39 });
    slider.simulate('keyUp', { keyCode: 39 });
    expect(wrapper.props().value).toEqual({ min: 2, max: 10 });

    slider.simulate('keyDown', { keyCode: 39 });
    slider.simulate('keyUp', { keyCode: 39 });
    expect(wrapper.props().value).toEqual({ min: 3, max: 10 });
  });

  it('should not respond to keyboard event when it is disabled', () => {
    const wrapper = mount(
      <RangeSlider
        {...defaultProps({ disabled: true, value: 2, onChange: (value: any) => wrapper.setProps({ value }) })}
      />
    );
    const slider = wrapper.find(`Slider [onKeyDown]`).first();

    slider.simulate('keyDown', { keyCode: 37 });
    slider.simulate('keyUp', { keyCode: 37 });
    expect(wrapper.props().value).toEqual(2);
  });

  it('should display the current value as a label', () => {
    const wrapper = mount(
      <RangeSlider
        {...defaultProps({ value: { min: 2, max: 10 } })}
      />
    );
    const label = wrapper.find('Slider Label').first();

    expect(label.text()).toEqual('2');
  });
});