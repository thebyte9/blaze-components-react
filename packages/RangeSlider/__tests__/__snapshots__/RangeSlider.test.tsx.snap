// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider component should be defined and renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    aria-disabled="false"
    class="input-range"
  >
    <span
      class="input-range__label input-range__label--min"
    >
      <span
        class="input-range__label-container"
      >
        0
      </span>
    </span>
    <div
      class="input-range__track input-range__track--background"
    >
      <div
        class="input-range__track input-range__track--active"
        style="left: 0%; width: 60%;"
      />
      <span
        class="input-range__slider-container"
        style="position: absolute; left: 60%;"
      >
        <span
          class="input-range__label input-range__label--value"
        >
          <span
            class="input-range__label-container"
          >
            12
          </span>
        </span>
        <div
          aria-valuemax="20"
          aria-valuemin="0"
          aria-valuenow="12"
          class="input-range__slider"
          draggable="false"
          role="slider"
          tabindex="0"
        />
      </span>
    </div>
    <span
      class="input-range__label input-range__label--max"
    >
      <span
        class="input-range__label-container"
      >
        20
      </span>
    </span>
  </div>
</DocumentFragment>
`;

exports[`RangeSlider component should display the current value as a label 1`] = `
<DocumentFragment>
  <div
    aria-disabled="false"
    class="input-range"
  >
    <span
      class="input-range__label input-range__label--min"
    >
      <span
        class="input-range__label-container"
      >
        0
      </span>
    </span>
    <div
      class="input-range__track input-range__track--background"
    >
      <div
        class="input-range__track input-range__track--active"
        style="left: 10%; width: 40%;"
      />
      <span
        class="input-range__slider-container"
        style="position: absolute; left: 10%;"
      >
        <span
          class="input-range__label input-range__label--value"
        >
          <span
            class="input-range__label-container"
          >
            2
          </span>
        </span>
        <div
          aria-valuemax="10"
          aria-valuemin="0"
          aria-valuenow="2"
          class="input-range__slider"
          draggable="false"
          role="slider"
          tabindex="0"
        />
      </span>
      <span
        class="input-range__slider-container"
        style="position: absolute; left: 50%;"
      >
        <span
          class="input-range__label input-range__label--value"
        >
          <span
            class="input-range__label-container"
          >
            10
          </span>
        </span>
        <div
          aria-valuemax="20"
          aria-valuemin="2"
          aria-valuenow="10"
          class="input-range__slider"
          draggable="false"
          role="slider"
          tabindex="0"
        />
      </span>
    </div>
    <span
      class="input-range__label input-range__label--max"
    >
      <span
        class="input-range__label-container"
      >
        20
      </span>
    </span>
  </div>
</DocumentFragment>
`;
